File: static/js/senter.js
$('#newAbbrev').keypress(function (e) {
    if (e.which == 13) {
       saveAbbrev();
       $('#newAbbrev').val('');
    }
  });

loadAbbreviations();

function showAbbreviations() {
    $('#btnShowAbbrev').hide();
    $('#btnHideAbbrev').show();
    $('#abbreviations').show(); 
}

function hideAbbreviations() {
    $('#btnShowAbbrev').show();
    $('#btnHideAbbrev').hide();
    $('#abbreviations').hide(); 
}

var abbreviations = []

function saveAbbrev() {

    $.ajax({
        type: 'POST',
        url: '/senter/abbrev/new',
        headers: {
            "Authorization": sessionStorage.getItem('simpligo.pln.jwt.key')
        },
        data: JSON.stringify({name: $('#newAbbrev').val()}),
        contentType: "application/json"
    }).done(function(data) {
        loadAbbreviations()
    }).fail(function(error) {
        alert( "Erro" );
    });

}

function loadAbbreviations() {
    $.ajax({
        type: 'GET',
        url: '/senter/abbrev/list',
        headers: {
            "Authorization": sessionStorage.getItem('simpligo.pln.jwt.key')
        }
    }).done(function(data) {
        var result = JSON.parse(data);
        var lista = "";
        abbreviations = [];
        result.list.forEach(item => {
            abbreviations.push(item.name);
            lista += item.name + "<span><a onclick='removeAbbrev(\"" + item.id + "\")'><i class='fa fa-trash-o inline-inner-button' data-toggle='tooltip' title='Excluir'></i> </a></span>"
        })
        $('#listaAbbrev').html(lista);
        
    }).fail(function(error) {
        alert( "Erro" );
    });
}

function removeAbbrev(abbrevKey) {
    $.ajax({
        type: 'DELETE',
        url: '/senter/abbrev/' + abbrevKey,
        headers: {
            "Authorization": sessionStorage.getItem('simpligo.pln.jwt.key')
        },
    }).done(function(data) {
        loadAbbreviations()
    }).fail(function(error) {
        alert( "Erro" );
    });

}

/* --------------------------------- */

var totalParagraphs;
var totalSentences;
var totalTokens;
var rawOutput = '';
var xmlOutput = '';
var jsonOutput = '';
var jsonOutputNoFormat = '';

function backToMenu() {
  loadMenu("/");
}

function split() {
    $('#results').show();

    var parsedText = splitText($('#rawContent').val());
  
    this.totalParagraphs = parsedText['totP'];
    this.totalSentences =  parsedText['totS'];
    this.totalTokens =  parsedText['totT'];
  
    this.rawOutput = this.outputToRaw(parsedText);
    this.xmlOutput = this.outputToXML(parsedText);
    this.jsonOutput = this.outputToJSON(parsedText);
    this.jsonOutputNoFormat = this.outputToJSONNoFormat(parsedText);
    
    $('#rawOutput').html(rawOutput);
    $('#xmlOutput').val(xmlOutput);
    $('#jsonOutput').val(jsonOutput);
    $('#jsonOutputNoFormat').val(jsonOutputNoFormat);

    $('#totalParagraphs').text(totalParagraphs);
    $('#totalSentences').text(totalSentences);
    $('#totalTokens').text(totalTokens);

}


function outputToRaw(parsedText) {
  var out = '';
  parsedText['paragraphs'].forEach(p => {
    p['sentences'].forEach(s => {
      out += 's' + s['idx'] + ' - ' + s['text'] + '<br/>';
    });
  });
  return out;
}

function outputToXML(parsedText) {
  var out = '';
  out = '<text>\n';
  parsedText['paragraphs'].forEach(p => {
    out += '  <p i=\"' + p['idx'] + '\">\n';
    p['sentences'].forEach(s => {
      out += '    <s i=\"' + s['idx'] + '\">' + s['text'] + "</s>\n";          
    });
    out += '  </p>\n';
  });
  out += '<totP>' + parsedText['totP'] + '</totP>\n'
  out += '<totS>' + parsedText['totS'] + '</totS>\n'
  out += '</text>';
  return out;
}

function outputToJSON(parsedText) {
  var out = '';
  out = '{\n  \"text\": [\n';
  parsedText['paragraphs'].forEach(p => {
    out += '            { \"i\": ' + p['idx'] + ',\n';
    out += '              \"p\":\"' + p['text'].replace(/"/g,'\\"') + '\",\n';
    out += '              \"ps\":[\n'
    p['sentences'].forEach(s => {
      out += '                   { \"i\": ' + s['idx'] + ', \"qtw\": ' + s['qtw'] + ', \"qtt\": ' + s['qtt'] + ', \"s\": \"' + s['text'].replace(/"/g,'\\"');
      out += '\",\n                     \"st\": [\n';
      s['tokens'].forEach(t => {
        out += '                             { \"i\": ' + t['idx'] + ', \"t\": \"' + t['token'] + '\"},\n';
      });
      out += '                     ]\n';
      out += '                   },\n';
    });
    out += '              ]\n            },\n';
  });
  out += '  ],\n';
  out += '  \"totP\":' + parsedText['totP'] + ',\n';
  out += '  \"totS\":' + parsedText['totS'] + ',\n';
  out += '  \"totT\":' + parsedText['totT'] + ',\n';
  out += '}';
  return out;
}

function outputToJSONNoFormat(parsedText) {
  var out = this.outputToJSON(parsedText);
  out = out.replace(/\n\s+/g, '');
  return out;
}

/* --------------------------------- */

/*
Rules:
1. Delimita-se uma sentença sempre que uma marca de nova linha (carriage return e line
feed) é encontrada, independentemente de um sinal de fim de sentença ter sido encontrado
anteriormente;
2. Não se delimitam sentenças dentro de aspas, parênteses, chaves e colchetes;
3. Delimita-se uma sentença quando os símbolos de interrogação (?) e exclamação (!) são
encontrados;
4. Delimita-se uma sentença quando o símbolo de ponto (.) é encontrado e este não é um
ponto de número decimal, não pertence a um símbolo de reticências (...), não faz parte de
endereços de e-mail e páginas da Internet e não é o ponto que segue uma abreviatura;
5. Delimita-se uma sentença quando uma letra maiúscula é encontrada após o sinal de
reticências ou de fecha-aspas.
(Pardo, 2006)
*/
function preProcesText(rawText) {
    var out = rawText;
    var match;

    // rule 2 - " { [ ( ) ] } "
    out = this.applyGroupRule(out, /"([^"]+?)"[^A-z]/g)
    out = this.applyGroupRule(out, /“(.+?)”/g)
    out = this.applyGroupRule(out, /\{(.+?)\}/g)
    out = this.applyGroupRule(out, /\[(.+?)\]/g)
    out = this.applyGroupRule(out, /\((.+?)\)/g)

    // rule 4 - abbreviations
    this.abbreviations.forEach(abbrev => {
        var abbrevNew = abbrev.replace(/\./g, '|dot|');
        var abbrevRe = abbrev.replace(/\./g, '\\.');
        var re = new RegExp(abbrevRe, "g");
        out = out.replace(re, abbrevNew);        
    });

    // rule 4 - internet
    var regexAddress = /(http|ftp|www|@)(.+?)\s/g;
    while (match = regexAddress.exec(out)) {
        var addressOld = match[2];
        var addressNew = addressOld.replace(/\./g, '|dot|');
        out = out.replace(addressOld, addressNew);
    }
    var regexEmail = /([A-z0-9_\-\.]+)@/g;
    while (match = regexEmail.exec(out)) {
        var emailOld = match[1];
        var emailNew = emailOld.replace(/\./g, '|dot|');
        out = out.replace(emailOld, emailNew);
    }

    //rule 4 - decimals
    out = out.replace(/([0-9]+)\.([0-9]+)/g, '$1|dot|$2');

    //rule 5 - quotes
    //out = out.replace(/"\s+([A-Z])/g, '\"|dot| |||$1'); // in texts well written this rule may be disabled.

    //rule 5 - reticences
    out = out.replace(/\.\.\.\s*([A-Z])/g, '|dot||dot||dot| |||$1');

    //rule 4 - reticences
    out = out.replace(/\.\.\./g, '|dot||dot||dot|');

    // rule 3
    out = out.replace(/\./g, '.|||');
    out = out.replace(/\?/g, '?|||');
    out = out.replace(/\!/g, '!|||');


    return out;
}

function applyGroupRule(rawText, regexGroup) {
    var match;
    while (match = regexGroup.exec(rawText)) {
        var sentenceOld = match[1];
        var sentenceNew = sentenceOld.replace(/\./g, '|gdot|');
        sentenceNew = sentenceNew.replace(/\?/g, '|gint|');
        sentenceNew = sentenceNew.replace(/\!/g, '|gexc|');
        rawText = rawText.replace(sentenceOld, sentenceNew);
    }
    return rawText;
}

function parseText(rawText) {
    var parsedText = {};
    var paragraphs = rawText.split("\n");

    var idxParagraphs = 0;
    var idxSentences = 0;
    var idxTokens = 0;
    var idxWords = 0;

    parsedText['paragraphs'] = [];

    paragraphs.forEach(p => {
        p = p.trim();
        if (p != '') {
        idxParagraphs++;
        var parsedParagraph = {"idx": idxParagraphs, "sentences": [], "text": p };
        var sentences = p.split("|||");
        sentences.forEach(s => {
            s = s.trim();
            if (s.length > 1) {
            idxSentences++;
            var parsedSentence = {"idx": idxSentences, "tokens": [], "text": s, "qtt": 0, "qtw": 0};
            var tokens = this.tokenizeText(s);
            var qtw = 0;
            var qtt = 0;
            tokens.forEach(t => {
                if (t.length > 0) {
                idxTokens++;
                qtt++;
                parsedSentence['tokens'].push({"idx": idxTokens, "token": t});
                if (t.length > 1 || '{[()]}.,"?!;:-\'#'.indexOf(t) < 0) {
                    qtw++;
                    idxWords++;
                }
                }
            });
            parsedSentence['qtt'] = qtt;
            parsedSentence['qtw'] = qtw;

            parsedParagraph['sentences'].push(parsedSentence);
            }
        });
        parsedText['paragraphs'].push(parsedParagraph);
        }
    });
    parsedText['totP'] = idxParagraphs;
    parsedText['totS'] = idxSentences;
    parsedText['totT'] = idxTokens;
    parsedText['totW'] = idxWords;

    return parsedText;
}

function postProcess(parsedText) {

    parsedText['paragraphs'].forEach(p => {
        p['text'] = this.punctuateBack(p['text']);
        p['sentences'].forEach(s => {
        s['text'] = this.punctuateBack(s['text']);
        s['tokens'].forEach(t => {
            t['token'] = this.punctuateBack(t['token']);
        });
        });
    });

    return parsedText;
}

function punctuateBack(text) {
    text = text.replace(/\|dot\|/g, '.');
    text = text.replace(/\|gdot\|/g, '.');
    text = text.replace(/\|gint\|/g, '?');
    text = text.replace(/\|gexc\|/g, '!');    
    text = text.replace(/\|hyp\|/g, '-');
    text = text.replace(/\|\|\|/g, '')
    return text;
}

function splitText(rawText) {

    rawText = this.preProcesText(rawText);

    var parsedText = this.parseText(rawText);    

    parsedText = this.postProcess(parsedText);

    return parsedText;
}

function tokenizeText(rawText) {
    rawText = rawText.replace(/([A-z]+)-([A-z]+)/g, "$1|hyp|$2");
    rawText = rawText.replace(/\|gdot\|/g, ".");
    rawText = rawText.replace(/\|gint\|/g, "?");
    rawText = rawText.replace(/\|gexc\|/g, "!");
    rawText = rawText.replace(/([\.\,"\(\)\[\]\{\}\?\!;:-]{1})/g, " $1 ");
    rawText = rawText.replace(/\s+/g, ' ');
    return rawText.split(' ');
}

File: tools/senter/senter.go
package senter

import (
	"encoding/json"
	"log"
	"regexp"
	"strings"
)

// func main() {
// 	a := `
// 	Muito interessante. A coisa funciona. O número 23.123, e o http://coisa.com. Isso mesmo.
// 	Também o sid.leal@gmail.com. Pois bem... Assim que aconteceu? Foi o Prof. João?
// 	Acho que foi! E ainda assim funciona. Ele disse: "Aqui nao pode ter quebra. Mesmo tendo ponto."
// 	E nem (aqui. Ignore isso.).
// 	Nascido em 230 A.C. com louvor.
// 	O Neil M. Ferguson apareceu naquela época.

// 	Ferguson disse: "Você precisa de uma combinação de estratégias. Nenhuma sozinha conseguiria prevenir com sucesso uma epidemia."`

// 	parsedText := ParseText(a)
// 	log.Println(parsedText)

// 	jsonText := ParseTextToJson(a)
// 	log.Println(jsonText)

// 	for _, p := range parsedText.Paragraphs {
// 		for _, s := range p.Sentences {
// 			log.Println(s.Text)
// 		}
// 	}

// }

type ParsedText struct {
	Paragraphs      []ParsedParagraph `json:"paragraphs"`
	TotalParagraphs int64             `json:"totp"`
	TotalSentences  int64             `json:"tots"`
	TotalTokens     int64             `json:"tott"`
	TotalWords      int64             `json:"totw"`
}

type ParsedParagraph struct {
	Idx          int64            `json:"idx"`
	Sentences    []ParsedSentence `json:"sentences"`
	Text         string           `json:"txt"`
	QtyTokens    int64            `json:"qtt"`
	QtyWords     int64            `json:"qtw"`
	QtySentences int64            `json:"qts"`
}

type ParsedSentence struct {
	Idx       int64         `json:"idx"`
	Tokens    []ParsedToken `json:"tokens"`
	Text      string        `json:"txt"`
	QtyTokens int64         `json:"qtt"`
	QtyWords  int64         `json:"qtw"`
}

type ParsedToken struct {
	Idx    int64  `json:"idx"`
	Token  string `json:"token"`
	IsWord int64  `json:"w"`
}

var abbreviations []string = []string{
	"Prof.", "A.C.", "Jr.", "a.C.",
}

func ParseText(rawText string) ParsedText {
	preText := preProcesText(rawText)
	processedText := processText(preText)
	parsedText := postProcess(processedText)
	return parsedText
}

func ParseTextToJson(rawText string) string {
	parsedText := ParseText(rawText)
	ret, err := json.Marshal(parsedText)
	if err != nil {
		log.Println(err)
	}
	return string(ret)
}

/*
Rules:
1. Delimita-se uma sentença sempre que uma marca de nova linha (carriage return e line
feed) é encontrada, independentemente de um sinal de fim de sentença ter sido encontrado
anteriormente;
2. Não se delimitam sentenças dentro de aspas, parênteses, chaves e colchetes;
3. Delimita-se uma sentença quando os símbolos de interrogação (?) e exclamação (!) são
encontrados;
4. Delimita-se uma sentença quando o símbolo de ponto (.) é encontrado e este não é um
ponto de número decimal, não pertence a um símbolo de reticências (...), não faz parte de
endereços de e-mail e páginas da Internet e não é o ponto que segue uma abreviatura;
5. Delimita-se uma sentença quando uma letra maiúscula é encontrada após o sinal de
reticências ou de fecha-aspas.
(Pardo, 2006)
6. Abreviações de nome próprio, espaçoMaisculaPonto
*/

func preProcesText(rawText string) string {
	out := rawText + "\n"

	// rule 2 - " { [ ( ) ] } "
	out = applyGroupRule(out, `"([^"]+?)"[^A-z]`)
	out = applyGroupRule(out, `“(.+?)”`)
	out = applyGroupRule(out, `\{(.+?)\}`)
	out = applyGroupRule(out, `\[(.+?)\]`)
	out = applyGroupRule(out, `\((.+?)\)`)

	// rule 4 - abbreviations
	for _, abbrev := range abbreviations {
		abbrevNew := strings.Replace(abbrev, ".", "|dot|", -1)
		out = strings.Replace(out, abbrev, abbrevNew, -1)
	}

	// rule 4 - internet
	regEx := regexp.MustCompile(`(http|ftp|www|@)(?P<dot>.+?)(,|\.)*\s`)
	matches := regEx.FindAllStringSubmatch(out, -1)
	for _, match := range matches {
		addressOld := match[2]
		addressNew := strings.Replace(addressOld, ".", "|dot|", -1)
		addressNew = strings.Replace(addressNew, ":", "|col|", -1)
		out = strings.Replace(out, addressOld, addressNew, -1)
	}

	regEx = regexp.MustCompile(`([A-z0-9_\-\.]+)@`)
	matches = regEx.FindAllStringSubmatch(out, -1)
	for _, match := range matches {
		emailOld := match[0]
		emailNew := strings.Replace(emailOld, ".", "|dot|", -1)
		out = strings.Replace(out, emailOld, emailNew, -1)
	}

	//rule 4 - decimals
	regEx = regexp.MustCompile(`([0-9]+)\.([0-9]+),([0-9]+)`)
	out = regEx.ReplaceAllString(out, `$1|dot|$2|virg|$3`)

	regEx = regexp.MustCompile(`([0-9]+),([0-9]+)`)
	out = regEx.ReplaceAllString(out, `$1|virg|$2`)

	regEx = regexp.MustCompile(`([0-9]+)\.([0-9]+)`)
	out = regEx.ReplaceAllString(out, `$1|dot|$2`)

	//rule 5 - quotes
	regEx = regexp.MustCompile(`"[\s\n]+?([A-Z])`)
	out = regEx.ReplaceAllString(out, `". $1`)

	//rule 5 - reticences
	regEx = regexp.MustCompile(`\.\.\.\s*([A-Z])`)
	out = regEx.ReplaceAllString(out, `|dot||dot||dot| |||$1`)

	//rule 4 - reticences
	regEx = regexp.MustCompile(`\.\.\.`)
	out = regEx.ReplaceAllString(out, `|dot||dot||dot|`)

	//rule 6 - nomes
	regEx = regexp.MustCompile(`(\s[A-Z]{1})\.(\s)`)
	out = regEx.ReplaceAllString(out, `$1|dot|$2`)

	// rule 3
	out = strings.Replace(out, ".", ".|||", -1)
	out = strings.Replace(out, "?", "?|||", -1)
	out = strings.Replace(out, "!", "!|||", -1)

	return out
}

func applyGroupRule(rawText string, regexGroup string) string {
	regEx := regexp.MustCompile(regexGroup)
	matches := regEx.FindAllStringSubmatch(rawText, -1)
	for _, match := range matches {
		sentenceOld := match[1]
		sentenceNew := strings.Replace(sentenceOld, ".", "|gdot|", -1)
		sentenceNew = strings.Replace(sentenceNew, "?", "|gint|", -1)
		sentenceNew = strings.Replace(sentenceNew, "!", "|gexc|", -1)
		rawText = strings.Replace(rawText, sentenceOld, sentenceNew, -1)
	}
	return rawText
}

func tokenizeText(rawText string) []string {
	regEx := regexp.MustCompile(`([A-z]+)-([A-z]+)`)
	rawText = regEx.ReplaceAllString(rawText, "$1|hyp|$2")

	regEx = regexp.MustCompile(`\|gdot\|`)
	rawText = regEx.ReplaceAllString(rawText, ".")

	regEx = regexp.MustCompile(`\|gint\|`)
	rawText = regEx.ReplaceAllString(rawText, "?")

	regEx = regexp.MustCompile(`\|gexc\|`)
	rawText = regEx.ReplaceAllString(rawText, "!")

	regEx = regexp.MustCompile(`([\.\,"\(\)\[\]\{\}\?\!;:-]{1})`)
	rawText = regEx.ReplaceAllString(rawText, "  $1 ")

	regEx = regexp.MustCompile(`\s+`)
	rawText = regEx.ReplaceAllString(rawText, " ")

	return strings.Split(rawText, " ")
}

func processText(rawText string) ParsedText {
	parsedText := ParsedText{}
	paragraphs := strings.Split(rawText, "\n")

	var idxParagraphs int64 = 0
	var idxSentences int64 = 0
	var idxTokens int64 = 0
	var idxWords int64 = 0

	parsedText.Paragraphs = []ParsedParagraph{}

	for _, p := range paragraphs {
		p = strings.TrimSpace(p)
		if p != "" {
			idxParagraphs++
			parsedParagraph := ParsedParagraph{}
			parsedParagraph.Idx = idxParagraphs
			parsedParagraph.Sentences = []ParsedSentence{}
			parsedParagraph.Text = p
			var qtwP int64 = 0
			var qttP int64 = 0
			sentences := strings.Split(p, "|||")
			for _, s := range sentences {
				s = strings.TrimSpace(s)
				if s != "" {
					idxSentences++
					parsedSentence := ParsedSentence{idxSentences, []ParsedToken{}, s, 0, 0}

					tokens := tokenizeText(s)
					var qtw int64 = 0
					var qtt int64 = 0
					for _, t := range tokens {
						if len(t) > 0 {
							idxTokens++
							qtt++
							qttP++
							token := ParsedToken{idxTokens, t, 0}
							if len(t) > 1 || strings.Index(`{[()]}.,"?!;:-'#&`, t) < 0 {
								qtw++
								qtwP++
								idxWords++
								token.IsWord = 1
							}
							parsedSentence.Tokens = append(parsedSentence.Tokens, token)
						}
					}
					parsedSentence.QtyTokens = qtt
					parsedSentence.QtyWords = qtw
					parsedParagraph.Sentences = append(parsedParagraph.Sentences, parsedSentence)
				}
			}
			parsedParagraph.QtyTokens = qttP
			parsedParagraph.QtyWords = qtwP
			parsedParagraph.QtySentences = int64(len(parsedParagraph.Sentences))
			parsedText.Paragraphs = append(parsedText.Paragraphs, parsedParagraph)
		}

	}
	parsedText.TotalParagraphs = idxParagraphs
	parsedText.TotalSentences = idxSentences
	parsedText.TotalTokens = idxTokens
	parsedText.TotalWords = idxWords

	return parsedText
}

func postProcess(parsedText ParsedText) ParsedText {

	for i, p := range parsedText.Paragraphs {
		parsedText.Paragraphs[i].Text = punctuateBack(p.Text)

		for j, s := range p.Sentences {
			parsedText.Paragraphs[i].Sentences[j].Text = punctuateBack(s.Text)
			for k, t := range s.Tokens {
				parsedText.Paragraphs[i].Sentences[j].Tokens[k].Token = punctuateBack(t.Token)
			}
		}
	}
	return parsedText
}

func punctuateBack(text string) string {
	text = strings.Replace(text, "|dot|", ".", -1)
	text = strings.Replace(text, "|int|", "?", -1)
	text = strings.Replace(text, "|exc|", "!", -1)
	text = strings.Replace(text, "|gdot|", ".", -1)
	text = strings.Replace(text, "|gint|", "?", -1)
	text = strings.Replace(text, "|gexc|", "!", -1)
	text = strings.Replace(text, "|hyp|", "-", -1)
	text = strings.Replace(text, "|col|", ":", -1)
	text = strings.Replace(text, "|||", "", -1)
	text = strings.Replace(text, "|virg|", ",", -1)
	return text
}
